stan/math/rev/core/vari.hpp

```diff
@@ -27,21 +58,29 @@ class var;
  * classes will store operand variables and propagate derivative
  * information via an implementation of chain().
  */
-class vari {
+template <typename T, typename = void>
+class vari_value;
+
+template <typename T>
+class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+    : public vari_base {
  private:
-  friend class var;
+  template <typename, typename>
+  friend class var_value;
 
  public:
+  using Scalar = T;
+  using value_type = Scalar;
```

Comment:
What's the logic on scalar vs. value type here?

Similarly for `var`.

I would expect:

`scalar_type_t<var_type<T>>` to be `var_type<T>`

`value_type_t<var_type<T>>` to be `T`

`value_type_t<vari_type<T>>` to be `T`

I don't know what I'd expect `scalar_type_t<vari_type<T>>` to be.

Anyway I just type out what I expect not to say things should be this way, but to clarify the question I am asking.

Reply:
@t4c1 had a good comment on this [here](https://github.com/stan-dev/math/issues/1913#issuecomment-634909816) that lines up with what you are thinking. I think value_type_t and scalar_type should still just go to `var_value<T>` and `vari_value<T>`. I'm still trying to munge this out

Reply:
Yeah that comment makes sense. And I also think @t4c1 is getting at the right thing that we kinda need to think about more levels of resolution.

Consider `std::complex<var_type<double>>`

We could do:

value_type as `std::complex<var_type<double>>`
scalar_type as `var_type<double>`

But then what is the `double` called? Cause surely we need access to that sometimes.

Also what is the scalar_type of:

`var_type<Eigen::MatrixXd>`?

Is it `Eigen::MatrixXd`? Or is it `double`?

Reply:
Somehow in the merge I'm getting a:

`vari_value<Eigen::Matrix<stan::math::var_value<double>, -1, -1>>`

Reply:
That very well may be from the `value_type` stuff I added in this PR.

I'm going to try out @t4c1's suggestion tmrw so that scalar_type and value_type stop at var_value since it feels like what we want from those are to look in containers then we will have something else for looking into the "value type" of scalars like `complex<T>` and `var_value<T>`

Reply:
So I think this is probably what we do now. At least it is one possibility (and I'm leaving off `std::vector` behavior):

```
value_type_t<double> -> double
scalar_type_t<double> -> double
value_type_t<std::complex<double>> -> double
scalar_type_t<std::complex<double>> -> std::complex<double>
value_type_t<Eigen::MatrixXd> -> double
scalar_type_t<Eigen::MatrixXd> -> double
```

We can think of value_type/scalar_type as autodiff agnostic:
```
value_type_t<var_type<double>> -> var_type<double>
scalar_type_t<var_type<double>> -> var_type<double>
value_type_t<std::complex<var_type<double>>> -> var_type<double>
scalar_type_t<std::complex<var_type<double>>> -> std::complex<var_type<double>>
value_type_t<var_type<Eigen::MatrixXd>> -> var_type<double>
scalar_type_t<var_type<Eigen::MatrixXd>> -> var_type<double>
value_type_t<Eigen::Matrix<var_type<double, -1, -1>> -> var_type<double>
scalar_type_t<Eigen::Matrix<var_type<double, -1, -1>> -> var_type<double>
```

Or we can think of value_type as something that goes into the autodiff (this seems more interesting just in that there are more differences which means it's probably more useful):
```
value_type_t<var_type<double>> -> double
scalar_type_t<var_type<double>> -> var_type<double>
value_type_t<std::complex<var_type<double>>> -> double
scalar_type_t<std::complex<var_type<double>>> -> std::complex<var_type<double>>
value_type_t<var_type<Eigen::MatrixXd>> -> double
scalar_type_t<var_type<Eigen::MatrixXd>> -> var_type<double>
value_type_t<Eigen::Matrix<var_type<double, -1, -1>> -> double
scalar_type_t<Eigen::Matrix<var_type<double, -1, -1>> -> var_type<double>
```

Anyway I think we need:
1. A way to extract the `T` from a `var_type<T>` or `vari_type<T>`.
2. And might as well have a way to get the `vari_type<T>` for a `var_type<T>` (this would be simple to implement with 1)

There's probably things I'm missing. Anyway I think tomorrow as I continue working on this I will just have to pick a standard and implement it. I wouldn't not be surprised if we end up out of sync, but I also don't expect to be able to write and agree on a full specification on what we need without more experimentation and being confused.

The adj_jac_apply changes in this pull are really gnarly too. Do you use that anywhere? If not, I might have to pull a Steve and break that out into it's own pull.

Reply:
> Also what is the scalar_type of:
var_type<Eigen::MatrixXd>?
Is it Eigen::MatrixXd? Or is it double?

Imo it should be var_type<double>.

>value_type_t<std::complex<double>> -> double
scalar_type_t<std::complex<double>> -> std::complex<double>

That is confusing. So far `scalar_type` was repeated ypplication of `value_type` untill a scalar remains. I think both should result in `std::complex<double>`. Than we make a separate trait for extracting real type from complex. 

> We can think of value_type/scalar_type as autodiff agnostic:

I aprove.

> Or we can think of value_type as something that goes into the autodiff (this seems more interesting just in that there are more differences which means it's probably more useful):

It might be more useful in the way that we need lesss traits to do everything. Bt it definitely is not easier to use ir understand.

Reply:
Apologies it took me a minute to respond this was confusing me a good bit and I had to play around with some code. 

> Imo it should be var_type<double>.

Oh? That's kind of a wild take to me, why so?

imo it feels like when we write specializations for `var_type<MatrixXd>` we'll want something like `require_var_vt<is_eigen, T>` no?

> That is confusing. So far scalar_type was equivalent to repeated application of value_type untill a scalar remains. I think both should result in std::complex<double>. Than we make a separate trait for extracting real type from complex.

>> We can think of value_type/scalar_type as autodiff agnostic:

> I approve.

It feels like the issue/confusion here is the conceptual vs mechanical application of the type traits. I setup a godbolt [here](https://godbolt.org/z/xHtvXz) that has all of our current value_type and scalar_type definitions and runs them over a bunch of different types (this includes the `var_value` specialization). tbh it's kind of a mess throwing in the `value_type` for `var`. But I have a few of these examples laid out below so feel free to cherry pick / modify the definitions of these things to show and tell.

[here](https://godbolt.org/z/2avqjB) is a version that instead treats everything equally. i.e. if a class has a `value_type` then `value_type_t` extracts it and `scalar_type_t` recursively goes through the `value_type`. imo this makes `scalar_type` kind of dumb, but they are consistent and now that we've cleaned these up we can make other more consistently defined type traits.

[this one](https://godbolt.org/z/zdHsKJ) Does a couple things

1. There's now `value_type` and `recursive_value_type` where `recursive_value_type` is just the previous definition of `scalar_type`. These do not respect our autodiff variables and will always extract the value_type.

2. `scalar_type` now extracts a class's `value_type` with a least upper bound (lub) of the set `{Arithmetic, var_value<T>, complex<T>, fvar<T>}` (i.e. it will not peek into these for a value_type)

3. `real_type` extracts a class's `value_type` with a lub in the set  `{Arithmetic, var_value<T>, fvar<T>}` (this one goes into complex)

4. They each have a `recursive_*` version that respects each type traits lub

I think this one covers all of our bases, going back to @t4c1's comment

> - extracting contained types from containers. scalar_type and value_type are in my opinion fine tools for that. The names of these two traits are also fine.

For containers I think the definitions of `value_type` and `recursive_value_type` make sense. We commonly are using these as `require_container_vt`, `require_eigen_vt`, etc. If we wanted to have a function that accepts `Container<complex<var<T>>>` where `T` is floating point could be done with something like

```cpp
// rrt is recursive_real_type
// is_var_fp is for vars that hold a floating point type
template <typename Container, 
  require_container_st<is_complex, Container>
  require_container_rrt<is_var_fp, Container>> 
```

> - extracting real type from a complex type. base_type_t combines that with what scalar_type does. Maybe we could use a trait that does just this. Maybe name it real_type?

Within a function body `value_type` can be used for this now. For templating we can use two requires like the above

> - extracting underlying type from a var, vari (or fvar). This is proposed to be added for var. For fvar value_type works that way. Maybe name it underlying type for going down one level and drop that thing with value_type? Going fully down to base (similar to scalar_type) could be named base_type if existing trait by that name gets renamed or removed.

- Again within a function body `value_type` is the tool, for checking a template for something like `var_value<Matrix<complex<double>>>` we have

```cpp
// rst is recursive scalar type
template <typename Var, require_var_vt<is_eigen, Var>,
  require_var_rst<is_complex, Var>>
```

> - converting other types to (at least) double. promote_args does that. Other traits combine this with other things.

If `promote_args_t` gets a `{var<double>, double, complex<double>}` should the return type be `complex<var<double>>`? And for `{var<Matrix<double>>, double, complex<double>}` what would be the return type?

Reply:
> Imo it should be var_type<double>

I agree with this now. I've been trying to use scalar like the mathematical sense. The goal being when I'm confused about what a scalar_type_t is I can just search "is X a scalar" and go by whatever the internet tells me.

And that's my argument for `var_type<MatrixXd>` not being a scalar type.

Thanks for that godbolt.

> They each have a recursive_* version that respects each type traits lub

I like the idea here -- having one version that looks down just one level, and one that goes all the way.

Reply:
> And that's my argument for var_type<MatrixXd> not being a scalar type.

Oh, yes! That's a very good point. That totally throws a shoehorn into my definitions

Reply:
> Thanks for that godbolt.

And yeah np! I think it's just easier to show and tell with this kind of stuff

Reply:
> And that's my argument for var_type<MatrixXd> not being a scalar type.

Though it's not a scalar type wouldn't these mostly just use `require_var_vt<is_eigen, Var>`? I think we could change my definition of `scalar_type` to be the lub of `{Arithmetic, var<Arithmetic>, fvar<T>, complex<T>`}` where `T` is a non-container type. Though then that's kind of weird because `recursive_scalar_type_t<var_value<Matrix<double>>` is going to give back `double`. idt that fits what we want does it?

Reply:
> idt that fits what we want does it

I'm having a hard time thinking through what we do/don't need.

I'm going to work on what is written for the multiplication operator:
```
template <typename T1, typename T2, require_all_var_value_t<T1, T2>* = nullptr>
inline auto operator*(const T1& a, const T2& b) {
  using vari1 = get_var_vari_value_t<T1>;
  using vari2 = get_var_vari_value_t<T2>;
  using scalar1_type = typename T1::value_type;
  using scalar2_type = typename T2::value_type;
  using mat_return
      = internal::mat_mul_return_type_t<scalar1_type, scalar2_type>;
  using multiply_type = internal::multiply_vari<mat_return, vari1, vari2>;
  return var_value<mat_return>{new multiply_type(a.vi_, b.vi_)};
}
```

I think across all combinations of prim/rev scalar/matrix/vector/row_vector there are like 196 different multiplications we might hit lol. Oh crap I just realized I forgot `std::complex` as well. Also that's not including mixes of `float`, `double`, `long double`. Can we just roll with double for now?

Anyway my goal is to vaguely make all those using statements and such go away.

Reply:
If we added a `vari_value_t` type trait to grab the `vari_type` from a `var_value` we could do something like

```cpp
template <typename T1, typename T2, require_all_var_t<T1, T2>* = nullptr>
inline auto operator*(const T1& a, const T2& b) {
  using vari1 = vari_value_t<T1>;
  using vari2 = vari_value_t<T2>;
  using scalar1_type = value_type_t<T1>;
  using scalar2_type = value_type_t<T2>;
  using mat_return = internal::mat_mul_return_t<scalar1_type, scalar2_type>;
  using multiply_type = internal::multiply_vari<mat_return, vari1, vari2>;
  return var_value<mat_return>{new multiply_type(a.vi_, b.vi_)};
}
```

Reply:
Though the reason for all the aliases is specifically because we have a bunch of different types. 

One Q though, Eigen doesn't allow for mixed type operations does it? i.e. `Matrix<float> * Matrix<double>` so that does cut out the allowed types that can pass into here (unless we want to decide to upcast those or something)

Reply:
Yeah, it might not be possible.

This doesn't work: https://godbolt.org/z/L6z9va

Reply:
Aight idt we should just do arbitrary upcasting and short/long stuff is not going to be user facing by the next release so let's not worry about it

But if everyone agrees (or has a different proposal) with v3 from my comment [here](https://github.com/stan-dev/math/pull/1915#discussion_r436290995) (pinging @bob-carpenter and @t4c1) then I think it would be a good idea to do those in a separate PR so this one doesn't blow up with a bunch of stuff

Reply:
> Anyway my goal is to vaguely make all those using statements and such go away.

I'm not following what you mean by this. Wouldn't that mean we would need a bunch of specializations?

Reply:
> Wouldn't that mean we would need a bunch of specializations?

Well I every time I try to think through it I get confused and just move code around, so maybe it's not possible to simplify, but that's what I'm gonna try.

Reply:
I'm not really following what you want to get rid of there. Like you could literally get rid of the usings by just calling them inline. Or you could just pass T1 and T2 to multiply_vari and do the deduction there like

```cpp
template <typename T1, typename T2, require_all_var_t<T1, T2>* = nullptr>
inline auto operator*(const T1& a, const T2& b) {
  using multiply_type = internal::multiply_vari<T1, T2>;
  // store the return type multiply_vari
  using mat_return = typename multiply_type::return_type;
  return var_value<mat_return>(new multiply_type(a.vi_, b.vi_)};
}
```

Reply:
> var_value<mat_return>
> internal::multiply_vari<T1, T2>

Both of these are constructors so you gotta pass the template arguments.

It'd be nice to not do that, cause the types should be deducible in the arguments. Presumably we have to define this somewhere, but doing it in every function would not be great.

This is connected to the multiple definitions of `operator*` and the multiple `multiply_vari` types.

I at least want less usings here.

Reply:
See the comment [here](https://github.com/stan-dev/math/pull/1884#issuecomment-640289620) I think we should keep convos here focused on just things in this particular PR

Reply:
> imo it feels like when we write specializations for var_type<MatrixXd> we'll want something like require_var_vt<is_eigen, T> no?

I'm not proposing to remove that - only to make a separate trait for it.

> But I have a few of these examples laid out below so feel free to cherry pick / modify the definitions of these things to show and tell.

Version 3 is the closest to what I had in mind. `value_type` still kinda does everything. I would delete that to encourage people to use specific traits. Or at least rename it so its name lists everything it does. Instead we need `base_type` for extracting from autodiff types. I would like `scalar type<var_value<Eigen::Matrix<double, -1, -1>>>` to be `var_value<double>`. `recursive_real_type` does not make much sense imo. `real_type<std::vector<std::complex<var_value<double> > >> = std::complex<var_value<double> >` seems like a bug. I think we should never nest matrices like `Eigen::Matrix<var_value<Eigen::Matrix<double, -1, -1> >, -1, -1, 0, -1, -1>`.

> And for {var<Matrix<double>>, double, complex<double>} what would be the return type?

var<Matrix<complex<double>>>

>I'm having a hard time thinking through what we do/don't need.

Having some combinations we dont need is a price I am willing to pay for trait system making more sense.

> If we added a vari_value_t type trait to grab the vari_type from a var_value we could do something like

Ugh, everything is a value. How about renaming this? Maybe `var2vari_t`?

> One Q though, Eigen doesn't allow for mixed type operations does it? i.e. Matrix<float> * Matrix<double> so that does cut out the allowed types that can pass into here (unless we want to decide to upcast those or something)

It does not. However there are some hacks in math that meke these work as long as one type is autodff type. This works only for element-wise operations. So no matrix mutiplications, solvers ... This should also be possible to do for non-autodiff types if we really wanted.

> I think it would be a good idea to do those in a separate PR so this one doesn't blow up with a bunch of stuff

I have some comments above. Other that those I agree.

> Or you could just pass T1 and T2 to multiply_vari and do the deduction there like

That makes sense.

> It'd be nice to not do that, cause the types should be deducible in the arguments. Presumably we have to define this somewhere, but doing it in every function would not be great.

Usual solution to this is to make a helper function that deduces templates, constructs the object and returns it. See kernel generator for examples.

Reply:
> Version 3 is the closest to what I had in mind. value_type still kinda does everything. I would delete that to encourage people to use specific traits. Or at least rename it so its name lists everything it does.

I'm gonna open up a seperate PR with this today. I like leaving value_type for now, but yeah renaming it seems like a good idea. Not sure what tho

> Instead we need base_type for extracting from autodiff types. I would like scalar type<var_value<Eigen::Matrix<double, -1, -1>>> to be var_value<double>. 

I'm not following the intuition here, why would we want this? You and Ben both seem to agree on this but I'm not totally seeing how that's the answer.

> recursive_real_type does not make much sense imo. 

Recursive `real_type` is for things like `std::vector<Matrix<complex<var<double>>>>` where we have a function that works on containers of containers containing a scalar type whose real type is a var<double>. Not sure if there is a use case for something that general/specific tho'

> real_type<std::vector<std::complex<var_value<double>>>> = std::complex<var_value<double> > seems like a bug. 

So note that without the `recursive_` part of the class it still just goes down one level so the recursive one would go down to `var_value<double>`. If you did

```cpp
real_type<real_type<std::vector<std::complex<var_value<double>>>>>
``` 

then it would get to var_value<double>

> I think we should never nest matrices like Eigen::Matrix<var_value<Eigen::Matrix<double, -1, -1> >, -1, -1, 0, -1, -1>.

I've only ever thought of one use case for this

```cpp
Eigen::SparseMatrix<var_value<Eigen::Matrix<double, -1, -1>>
```

I'm not sure if that would ever be a good idea, but it would let you have a sparse matrix composed of blocks of dense matrices. If we did want to stop that tho' is there a way to? Like can we disallow that by doing something clever in Eigen?

Reply:
This does technically work, though tbh if I ever saw this I'd almost always assume it was an error

```cpp
    Eigen::Matrix<Eigen::Matrix<double, -1, -1>, -1, -1> A(2, 2);
    for (int i = 0; i < A.size(); i++) {
        A(i) = Eigen::Matrix<double, -1, -1>::Random(2, 2);
    }
    Eigen::Matrix<double, -1, -1> B = A(1) * A(2);
```

Reply:
> I like leaving value_type for now

I would really prefere if we eventually remove that. Untill than it will be used everywhere even where other new traits would be better suited.

> Recursive real_type is for things like std::vector<Matrix<complex<var<double>>>> where we have a function that works on containers of containers containing a scalar type whose real type is a var. Not sure if there is a use case for something that general/specific tho'

Real type of that should be `std::vector<Matrix<var<double>>>`, regardless of it being recursive or not. Real type should just remove complex if it is anywhere in given type.

> I'm not following the intuition here, why would we want this? You and Ben both seem to agree on this but I'm not totally seeing how that's the answer.

Example use case: iterating over elements of a matrix or vector - single element is of type as determined by scalar_type.

Reply:
> So note that without the recursive_ part of the class it still just goes down one level so the recursive one would go down to var_value<double>. If you did

Yeah, but it should not remove `std::vector` or `Eigen::Matrix`.

> Like can we disallow that by doing something clever in Eigen?

I don't think so. But we also don't need to support that anywhere or even think about it.

Reply:
> I would really prefere if we eventually remove that. Untill than it will be used everywhere even where other new traits would be better suited.

Thinking about this more I'm agreeing with it. Lemme see what I can brew up today

> Real type of that should be `std::vector<Matrix<var<double>>>`, regardless of it being recursive or not. Real type should just remove complex if it is anywhere in given type.

The goal of those type traits is to be specializations of `value_type`, essentially `value_type` but they stop at a given bound. It sounds like what you are looking for is something that slices (cuts, subsets, modifies?) a type through exclusion of a certain type?

> I don't think so. But we also don't need to support that anywhere or even think about it.

I concur

Reply:
> The goal of those type traits is to be specializations of value_type

Here we disagree. Real type of a matrix being scalar does not make that much sense. We can make a bit deeper changes if resulting traits are consequentially more sensible.